wrap_error.go

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zap
+
+import (
+	"go.uber.org/zap/zapcore"
+)
+
+type errorWithFields struct {
+	err    error
+	fields []Field
+}
+
+// WrapError returns an error that will log the provided fields if the error
+// is logged using `Error`.
+func WrapError(err error, fields ...Field) error {
+	return &errorWithFields{
```

Comment:
```suggestion
// WrapError returns an error that, when logged with [Error] or [NamedError],
// will add the specified fields to the log entry.
//
// For example, given:
//
// 	func NewUser(uid string) (*User, error) {
//		if !isValid(uid) {
//			return nil, zap.WrapError(ErrInvalidUserID, zap.String("uid", uid))
//		}
//		// ...
//	}
//
// The following will log the error message, as well as the 'uid' field.
//
//	u, err := NewUser(uid)
//	if err != nil {
//		log.Warn("Unable to create user", zap.Error(err))
//	}
func WrapError(err error, fields ...Field) error {
	return &errorWithFields{
```

(I wanted to clarify that 'Error' refers to the field constructor, not the Logger.Error method and given that it's a new thing, more details on what it does and how to use it.)